File: LintCode_Python/twoSum.py
# Given an array of integers, find two numbers such that they add up to a specific
# target number.

# The function twoSum should return indices of the two numbers such that they add
# up to the target, where index1 must be less than index2. Please note that your
#  returned answers (both index1 and index2) are NOT zero-based.

# Have you met this question in a real interview? Yes
# Example
# numbers=[2, 7, 11, 15], target=9

# return [1, 2]

# Note
# You may assume that each input would have exactly one solution

# Challenge
# Either of the following solutions are acceptable:

# O(n) Space, O(nlogn) Time
# O(n) Space, O(n) Time


def twoSum(numbers, target):
    if not numbers or len(numbers) == 1:
        return None
    d = dict()
    for i in range(len(numbers)):
        d[numbers[i]] = i
    print d
    result = []
    for i in range(len(numbers)):
        if target - numbers[i] in d:
            result.append(i + 1)
            result.append(d[target - numbers[i]] + 1)
            return result
    return result


    # d = dict()
    # d[0] = -1
    # sums = 0
    # result = []
    # for i in range(len(numbers)):
    #     sums += numbers[i]
    #     if sums - target in d:
    #         result.append(d[sums - target] + 2)
    #         result.append(i + 1)
    #         if result[0] == result[1]:
    #             continue
    #         return result
    #     else:
    #         d[sums] = i
    #     print d
    # return result


# print twoSum([2, 7 ,11, 15], 9)
# print twoSum([1, 0, -1], -1)
print twoSum([1, 0, -1], 0)

File: LeetCode_Java/src/TwoSum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;


/**
 * Given an array of integers, return indices of the two numbers such that they add up to a specific target.
 * You may assume that each input would have exactly one solution.
 * Example:
 * Given nums = [2, 7, 11, 15], target = 9,
 * Because nums[0] + nums[1] = 2 + 7 = 9,
 * return [0, 1].
 * 
 * Tags: Array, HashTable
 * @author wendi
 *
 */
public class TwoSum {
	
//	/**
//	 * Method2: If given array is sorted, using two pointers
//	 * @param int[] nums, int target
//	 * @return int[]
//	 * Time: O(n)
//	 * Space: O(1)
//	 */
//	public int[] twoSumI(int[] nums, int target) {
//		if (nums == null || nums.length < 2) return new int[] {-1, -1};	
//		Arrays.sort(nums);
//		int left = 0;
//		int right = nums.length - 1;
//		while (left < right) {
//			if (nums[left] + nums[right] == target) return new int[] {left, right};
//			else if (nums[left] + nums[right] < target) left++;
//			else right--;
//		}
//		return new int[] {-1, -1};
//	}
	
	
	/**
	 * Method: one pass hashMap
	 * @param int[] nums, int target
	 * @return int[]
	 * Time: O(n)
	 * Space: O(n)
	 */
	public int[] twoSum(int[] nums, int target) {
		if (nums == null || nums.length < 2) return new int[] {-1, -1};
		Map<Integer, Integer> hash = new HashMap<>();
		for (int i = 0; i < nums.length; i++) {
			if (hash.containsKey(target - nums[i])) {
				return new int[] {hash.get(target - nums[i]), i};
			}
			else {
				hash.put(nums[i], i);
			}
		}
		return null;
	}
 
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		TwoSum result = new TwoSum();
		System.out.println(Arrays.toString(result.twoSum(new int[] {2, 7, 11, 15}, 9)));
		System.out.println(Arrays.toString(result.twoSum(new int[] {3, 5, 3}, 6)));
//		System.out.println(Arrays.toString(result.twoSumI(new int[] {2, 7, 11, 15}, 9)));
//		System.out.println(Arrays.toString(result.twoSumI(new int[] {3, 3, 5}, 6)));
	}

}
